File: src/main/java/org/tobi29/scapes/engine/graphics/TextureManager.java
/*
 * Copyright 2012-2015 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.tobi29.scapes.engine.graphics;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.tobi29.scapes.engine.ScapesEngine;
import org.tobi29.scapes.engine.utils.graphics.PNG;
import org.tobi29.scapes.engine.utils.io.ReadableByteStream;
import org.tobi29.scapes.engine.utils.io.filesystem.FileSystemContainer;
import org.tobi29.scapes.engine.utils.io.filesystem.ReadSource;

import java.io.IOException;
import java.io.InputStream;
import java.util.Map;
import java.util.Properties;
import java.util.concurrent.ConcurrentHashMap;

public class TextureManager {
    private static final Logger LOGGER =
            LoggerFactory.getLogger(TextureManager.class);
    private final ScapesEngine engine;
    private final Map<String, Texture> cache = new ConcurrentHashMap<>();

    public TextureManager(ScapesEngine engine) {
        this.engine = engine;
    }

    public void bind(String asset, GL gl) {
        if (asset == null) {
            unbind(gl);
        } else if (asset.isEmpty()) {
            unbind(gl);
        } else {
            bind(get(asset), gl);
        }
    }

    public void bind(Texture texture, GL gl) {
        if (texture == null) {
            unbind(gl);
        } else {
            texture.bind(gl);
        }
    }

    public Texture get(String asset) {
        Texture texture = cache.get(asset);
        if (texture == null) {
            texture = load(asset);
        }
        return texture;
    }

    public Texture empty() {
        return engine.graphics().textureEmpty();
    }

    private Texture load(String asset) {
        try {
            Properties properties = new Properties();
            FileSystemContainer files = engine.files();
            ReadSource imageResource = files.get(asset + ".png");
            ReadSource propertiesResource = files.get(asset + ".properties");
            if (propertiesResource.exists()) {
                try (InputStream streamIn = propertiesResource.readIO()) {
                    properties.load(streamIn);
                }
            }
            Texture texture = imageResource
                    .readReturn(stream -> texture(stream, properties));
            cache.put(asset, texture);
            return texture;
        } catch (IOException e) {
            LOGGER.error("Failed to load texture from: {} ({})", asset,
                    e.getMessage());
        }
        return empty();
    }

    private Texture texture(ReadableByteStream stream, Properties properties)
            throws IOException {
        return engine.graphics()
                .createTexture(PNG.decode(stream, engine::allocate),
                        Integer.valueOf(properties.getProperty("Mipmaps", "4")),
                        TextureFilter.get(properties
                                .getProperty("MinFilter", "Nearest")),
                        TextureFilter.get(properties
                                .getProperty("MagFilter", "Nearest")),
                        TextureWrap
                                .get(properties.getProperty("WrapS", "Repeat")),
                        TextureWrap.get(properties
                                .getProperty("WrapT", "Repeat")));
    }

    public void unbind(GL gl) {
        empty().bind(gl);
    }

    public void clearCache() {
        cache.clear();
    }
}

File: src/main/kotlin/org/tobi29/scapes/engine/graphics/TextureManager.kt
/*
 * Copyright 2012-2016 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.engine.graphics

import java8.util.Maps
import mu.KLogging
import org.tobi29.scapes.engine.ScapesEngine
import org.tobi29.scapes.engine.resource.Resource
import org.tobi29.scapes.engine.utils.graphics.decodePNG
import org.tobi29.scapes.engine.utils.io.ReadableByteStream
import java.util.*

class TextureManager(private val engine: ScapesEngine) {
    private val cache = WeakHashMap<String, Resource<Texture>>()

    @Synchronized
    operator fun get(asset: String): Resource<Texture> {
        return Maps.computeIfAbsent(cache, asset) {
            load(asset)
        }
    }

    fun empty(): Texture {
        return engine.graphics.textureEmpty()
    }

    private fun load(asset: String): Resource<Texture> {
        return engine.resources.load({
            empty()
        }) res@ {
            val properties = Properties()
            val files = engine.files
            val imageResource = files[asset + ".png"]
            val propertiesResource = files[asset + ".properties"]
            if (propertiesResource.exists()) {
                propertiesResource.readIO().use { streamIn ->
                    properties.load(streamIn)
                }
            }
            imageResource.read { texture(it, properties) }
        }
    }

    private fun texture(stream: ReadableByteStream,
                        properties: Properties): Texture {
        return engine.graphics.createTexture(decodePNG(stream,
                { engine.allocate(it) }),
                properties.getProperty("Mipmaps", "4").toInt(),
                TextureFilter[properties.getProperty("MinFilter", "Nearest")],
                TextureFilter[properties.getProperty("MagFilter", "Nearest")],
                TextureWrap[properties.getProperty("WrapS", "Repeat")],
                TextureWrap[properties.getProperty("WrapT", "Repeat")])
    }

    fun unbind(gl: GL) {
        empty().bind(gl)
    }

    @Synchronized
    fun clearCache() {
        cache.clear()
    }

    companion object : KLogging()
}
